### PR TITLE
Add Read-Write Splitting Plugin

### PR DESCRIPTION
**Pull request by congoamz**
### Summary

Implement Read-Write Splitting Plugin

### Additional Reviewers

@sergiyv-bitquill 
@karenc-bq

### CODE REVIEW

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+  JdbcConnection getWriterConnection() {
+    return this.writerConnection;
+  }
+
+  JdbcConnection getReaderConnection() {
+    return this.readerConnection;
+  }
+
+  private void logTopology() {
+    StringBuilder msg = new StringBuilder();
```
Comment (davecramer) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java:596]: would think we would want to put a guard around this. If logger level less than trace no sense in creating the string

File: src/main/resources/com/mysql/cj/LocalizedErrorMessages.properties [hunk 2]
```diff
@@ -1100,4 +1094,42 @@ ClusterAwareReaderFailoverHandler.2=[ClusterAwareReaderFailoverHandler] Connecte
 ClusterAwareReaderFailoverHandler.3=[ClusterAwareReaderFailoverHandler] Trying to connect to reader [{0,number,#}] ''{1}''
 ClusterAwareReaderFailoverHandler.4=[ClusterAwareReaderFailoverHandler] Connected to reader [{0,number,#}] ''{1}''
 ClusterAwareReaderFailoverHandler.5=[ClusterAwareReaderFailoverHandler] Failed to connect to reader [{0,number,#}] ''{1}''
-ClusterAwareReaderFailoverHandler.6=[ClusterAwareReaderFailoverHandler] {0} was called with an invalid (null or empty) topology
\ No newline at end of file
+
+ReadWriteSplittingPlugin.1=[ReadWriteSplittingPlugin] An error occurred while trying to switch to a reader connection.
```
Comment (davecramer) [inline src/main/resources/com/mysql/cj/LocalizedErrorMessages.properties:1098]: I think more descriptive names are better for readability in the code.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java (hunk 1 above)
Comment (congoamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java:596]: Good point, thanks Dave

File: src/main/resources/com/mysql/cj/LocalizedErrorMessages.properties (hunk 2 above)
Comment (congoamz) [reply src/main/resources/com/mysql/cj/LocalizedErrorMessages.properties:1098]: Good idea, will do

**Comment by samcrop** [comment]
Is anyone actively working on this? We still can't use this driver if it doesn't have read/write splitting. There's currently no driver that provides full functionality with Aurora and Java which seems crazy considering we're paying for the product. I would very much appreciate an update on this. Thanks.

**Comment by hsuamz** [comment]
> Is anyone actively working on this? We still can't use this driver if it doesn't have read/write splitting. There's currently no driver that provides full functionality with Aurora and Java which seems crazy considering we're paying for the product. I would very much appreciate an update on this. Thanks.

Thanks for reaching out @samcrop. We wanted to share more visibility into this. We've released a new driver here: https://github.com/awslabs/aws-advanced-jdbc-wrapper. The AWS JDBC Driver has been redesigned as a JDBC wrapper that is complementary to and extends the functionality of an existing JDBC driver to help an application take advantage of the features of clustered databases such as Amazon Aurora. The AWS JDBC Driver does not implement connectivity on its own to any database, but will enable support of AWS and Aurora functionalities on top of an underlying JDBC driver of the user's choice.

The AWS JDBC Driver is meant to work with any JDBC driver. Currently, the AWS JDBC Driver has been validated to support the [PostgreSQL JDBC Driver](https://github.com/pgjdbc/pgjdbc). Support for the [MySQL JDBC Driver](https://github.com/mysql/mysql-connector-j) and the [MariaDB JDBC Driver](https://github.com/mariadb-corporation/mariadb-connector-j) is being evaluated.

With that being said, we are planning to release Read / Write splitting for the JDBC wrapper driver. We are currently running through testing with MySQL and MariaDB before we claim official support with the JDBC wrapper driver and will encourage customers to move to that driver once ready. We are providing support for this driver, but will be saving new functionality for the JDBC wrapper driver.

We will share more concrete dates once available. Please let us know if you have any questions.

**Comment by samcrop** [comment]
Hi Matthew, thanks for the update, that does sound promising. Still no read/write splitting yet though I see. I hope this can be released fairly quickly as without it it seriously reduces the usefulness of our read replicas.

**Comment by gazal-k** [comment]
It looks like there are commits related to read write splitting plugin and compatibility with MySQL & MariaDB drivers. Looks very promising indeed.

**Comment by samcrop** [comment]
@hsuamz Could I please get an update on this? This is the JDBC driver recommended in all the AWS docs yet it doesn't support all the Aurora features that we're paying for (specifically read-write splitting) and thus we're paying for a product that we cannot fully utilise.

The newer advanced jdbc wrapper doesn't even support MySQL, let alone read/write splitting in production.
Thanks.

**Comment by hsuamz** [comment]
@samcrop - Read/Write Splitting is supported in the [aws-advanced-jdbc-wrapper](https://github.com/awslabs/aws-advanced-jdbc-wrapper), but it has been released as an experimental feature. The current implementation follows a similar approach to what we see in the Community MySQL JDBC driver. As for MySQL support, the `aws-advanced-jdbc-wrapper` is compatible with MySQL, but we have not made an official release. If you wanted to give the latest release a try, we’d appreciate any feedback given. Both these features will have an official release and we’ll share scheduling once confirmed.

Thanks,

Matthew

**Comment by gazal-k** [comment]
Would this driver be sunset in favour of the wrapper?

**Comment by hsuamz** [comment]
> Would this driver be sunset in favour of the wrapper?

@gazal-k : We are working to certify the `aws-advanced-jdbc-wrapper` for Aurora MySQL. Our development work is focused on that wrapper driver. We continue to support the driver contained in this repo, though.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java (hunk 1 above)
Comment (davecramer) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java:600]: This should have been fixed before closing this PR

Comment (hsuamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java:600]: This PR was closed as Read / Write Splitting has been released with the `aws-advanced-jdbc-wrapper`. As for MySQL support, the `aws-advanced-jdbc-wrapper` is compatible with MySQL and we are currently working towards an official MySQL release in the future.

Comment (davecramer) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java:600]: Yes, I figured that out too late.

**Pull request closed without merging**